File: backend/api/routers/ambp.py
"""
AutumnPlot Meteorological Binary Protocol (AMBP) v1.0
Author: Your Team
"""

import struct
import numpy as np
from typing import List, Tuple

# ==============================
# Protocol Constants
# ==============================

MAGIC = b"AMBP"
VERSION = 1

# Message Types
MSG_GRID = 1
MSG_POINT = 2
MSG_GEOMETRY = 3

# DType IDs
DTYPE_INT16 = 1
DTYPE_INT32 = 2
DTYPE_FLOAT32 = 3

# Geometry Types
GEOM_POLYLINE = 1
GEOM_POLYGON = 2

# ==============================
# Utility: Scale Packing
# ==============================

def auto_pack_int16(data: np.ndarray):
    """
    Automatically compute scale/offset and pack float array into int16.
    """
    if not np.isfinite(data).any():
        raise ValueError("Data contains no finite values.")

    data_min = np.nanmin(data)
    data_max = np.nanmax(data)

    scale = (data_max - data_min) / 65530.0 if data_max != data_min else 1.0
    offset = data_min

    packed = np.full(data.shape, -32768, dtype=np.int16)
    mask = np.isfinite(data)
    packed[mask] = np.round((data[mask] - offset) / scale).astype(np.int16)

    return packed, scale, offset


# ==============================
# Packet Framing
# ==============================

def _frame_packet(message_type: int, header_block: bytes, data_block: bytes) -> bytes:
    flags = 0
    header_length = len(header_block)

    return (
        MAGIC +
        struct.pack("<BBH", VERSION, message_type, flags) +
        struct.pack("<I", header_length) +
        header_block +
        data_block
    )


# ==============================
# Grid Message
# ==============================

def encode_grid(
    array: np.ndarray,
    variable_id: int,
    units_id: int,
    projection_id: int,
    bbox: Tuple[float, float, float, float],
    pack: bool = True
) -> bytes:
    """
    Encode a 2D grid into AMBP format.
    """

    if array.ndim != 2:
        raise ValueError("Grid must be 2D.")

    ny, nx = array.shape

    if pack:
        packed, scale, offset = auto_pack_int16(array.astype(np.float32))
        dtype_id = DTYPE_INT16
        data_block = packed.tobytes(order="C")
        missing = -32768
    else:
        scale = 1.0
        offset = 0.0
        dtype_id = DTYPE_FLOAT32
        packed = array.astype(np.float32)
        data_block = packed.tobytes(order="C")
        missing = -9999

    minlon, minlat, maxlon, maxlat = bbox

    header = struct.pack(
        "<IIffhBBBBffff",
        nx,
        ny,
        scale,
        offset,
        missing,
        dtype_id,
        variable_id,
        units_id,
        projection_id,
        minlon,
        minlat,
        maxlon,
        maxlat,
    )

    header = header.ljust(64, b"\x00")  # pad to 64 bytes

    return _frame_packet(MSG_GRID, header, data_block)


# ==============================
# Point Message
# ==============================

def encode_points(
    fields: List[Tuple[int, np.ndarray, int, float, float]]
) -> bytes:
    """
    fields: List of tuples:
        (field_id, array, dtype_id, scale, offset)
    """

    if not fields:
        raise ValueError("No fields provided.")

    count = len(fields[0][1])

    for _, arr, _, _, _ in fields:
        if len(arr) != count:
            raise ValueError("All point fields must have same length.")

    header = struct.pack("<IB3s", count, len(fields), b"\x00\x00\x00")

    field_descriptors = b""
    data_block = b""

    for field_id, arr, dtype_id, scale, offset in fields:
        field_descriptors += struct.pack(
            "<BBHff",
            field_id,
            dtype_id,
            0,
            scale,
            offset
        )

        if dtype_id == DTYPE_INT16:
            data_block += arr.astype(np.int16).tobytes()
        elif dtype_id == DTYPE_INT32:
            data_block += arr.astype(np.int32).tobytes()
        elif dtype_id == DTYPE_FLOAT32:
            data_block += arr.astype(np.float32).tobytes()
        else:
            raise ValueError("Unsupported dtype.")

    header_block = header + field_descriptors

    return _frame_packet(MSG_POINT, header_block, data_block)


# ==============================
# Geometry Message
# ==============================

def encode_geometries(
    geometries: List[Tuple[int, np.ndarray]]
) -> bytes:
    """
    geometries: List of tuples:
        (geometry_type, vertices Nx2 float array in degrees)
    """

    header = struct.pack("<I", len(geometries))

    data_block = b""

    for geom_type, vertices in geometries:
        if vertices.shape[1] != 2:
            raise ValueError("Vertices must be Nx2 array.")

        vertex_count = vertices.shape[0]

        # Scale to microdegrees
        scaled = np.round(vertices * 1e6).astype(np.int32)

        data_block += struct.pack("<B3sI", geom_type, b"\x00\x00\x00", vertex_count)
        data_block += scaled.tobytes()

    return _frame_packet(MSG_GEOMETRY, header, data_block)

File: frontend/src/decoders/ambp.js
/*
 * AutumnPlot Meteorological Binary Protocol (AMBP) v1.0 Decoder
 */

const AMBP_MAGIC = "AMBP";
const AMBP_VERSION = 1;

// Message Types
const MSG_GRID = 1;
const MSG_POINT = 2;
const MSG_GEOMETRY = 3;

// DTypes
const DTYPE_INT16 = 1;
const DTYPE_INT32 = 2;
const DTYPE_FLOAT32 = 3;

// Geometry Types
const GEOM_POLYLINE = 1;
const GEOM_POLYGON = 2;


// ==============================
// Public API
// ==============================

export function decodeAMBP(arrayBuffer) {
    const view = new DataView(arrayBuffer);

    // Validate magic
    const magic = readString(view, 0, 4);
    if (magic !== AMBP_MAGIC) {
        throw new Error("Invalid AMBP magic header");
    }

    const version = view.getUint8(4);
    if (version !== AMBP_VERSION) {
        throw new Error(`Unsupported AMBP version: ${version}`);
    }

    const messageType = view.getUint8(5);
    const flags = view.getUint16(6, true);
    const headerLength = view.getUint32(8, true);

    const headerOffset = 12;
    const dataOffset = headerOffset + headerLength;

    switch (messageType) {
        case MSG_GRID:
            return decodeGrid(arrayBuffer, headerOffset, dataOffset);
        case MSG_POINT:
            return decodePoints(arrayBuffer, headerOffset, dataOffset);
        case MSG_GEOMETRY:
            return decodeGeometry(arrayBuffer, headerOffset, dataOffset);
        default:
            throw new Error(`Unknown AMBP message type: ${messageType}`);
    }
}


// ==============================
// Grid Decoder
// ==============================

function decodeGrid(buffer, headerOffset, dataOffset) {
    const view = new DataView(buffer);

    let o = headerOffset;

    const nx = view.getUint32(o, true); o += 4;
    const ny = view.getUint32(o, true); o += 4;
    const scale = view.getFloat32(o, true); o += 4;
    const offset = view.getFloat32(o, true); o += 4;
    const missing = view.getInt16(o, true); o += 2;

    const dtype = view.getUint8(o); o += 1;
    const variableId = view.getUint8(o); o += 1;
    const unitsId = view.getUint8(o); o += 1;
    const projectionId = view.getUint8(o); o += 1;

    const bbox = [
        view.getFloat32(o, true),
        view.getFloat32(o + 4, true),
        view.getFloat32(o + 8, true),
        view.getFloat32(o + 12, true)
    ];

    const count = nx * ny;

    let data;

    if (dtype === DTYPE_INT16) {
        data = new Int16Array(buffer, dataOffset, count);
    } else if (dtype === DTYPE_INT32) {
        data = new Int32Array(buffer, dataOffset, count);
    } else if (dtype === DTYPE_FLOAT32) {
        data = new Float32Array(buffer, dataOffset, count);
    } else {
        throw new Error("Unsupported grid dtype");
    }

    return {
        type: "grid",
        nx,
        ny,
        scale,
        offset,
        missing,
        variableId,
        unitsId,
        projectionId,
        bbox,
        data
    };
}


// ==============================
// Point Decoder
// ==============================

function decodePoints(buffer, headerOffset, dataOffset) {
    const view = new DataView(buffer);

    let o = headerOffset;

    const count = view.getUint32(o, true); o += 4;
    const fieldCount = view.getUint8(o); o += 1;
    o += 3; // reserved

    const fields = [];

    for (let i = 0; i < fieldCount; i++) {
        const fieldId = view.getUint8(o); o += 1;
        const dtype = view.getUint8(o); o += 1;
        o += 2; // reserved
        const scale = view.getFloat32(o, true); o += 4;
        const offset = view.getFloat32(o, true); o += 4;

        fields.push({ fieldId, dtype, scale, offset });
    }

    const result = {};
    let dataPtr = dataOffset;

    for (let f of fields) {
        let arr;

        if (f.dtype === DTYPE_INT16) {
            arr = new Int16Array(buffer, dataPtr, count);
            dataPtr += count * 2;
        } else if (f.dtype === DTYPE_INT32) {
            arr = new Int32Array(buffer, dataPtr, count);
            dataPtr += count * 4;
        } else if (f.dtype === DTYPE_FLOAT32) {
            arr = new Float32Array(buffer, dataPtr, count);
            dataPtr += count * 4;
        } else {
            throw new Error("Unsupported point dtype");
        }

        result[f.fieldId] = {
            data: arr,
            scale: f.scale,
            offset: f.offset
        };
    }

    return {
        type: "point",
        count,
        fields: result
    };
}


// ==============================
// Geometry Decoder
// ==============================

function decodeGeometry(buffer, headerOffset, dataOffset) {
    const view = new DataView(buffer);

    let o = headerOffset;
    const featureCount = view.getUint32(o, true);

    let dataPtr = dataOffset;

    const features = [];

    for (let i = 0; i < featureCount; i++) {
        const geomType = view.getUint8(dataPtr); dataPtr += 1;
        dataPtr += 3; // reserved

        const vertexCount = view.getUint32(dataPtr, true);
        dataPtr += 4;

        const coords = new Int32Array(buffer, dataPtr, vertexCount * 2);
        dataPtr += vertexCount * 2 * 4;

        features.push({
            geometryType: geomType,
            vertexCount,
            coords,
            scale: 1e-6
        });
    }

    return {
        type: "geometry",
        featureCount,
        features
    };
}


// ==============================
// Helpers
// ==============================

function readString(view, offset, length) {
    let s = "";
    for (let i = 0; i < length; i++) {
        s += String.fromCharCode(view.getUint8(offset + i));
    }
    return s;
}